Issue #441. Add methods for semantic navigation

sergisiso opened the pull request:
- [x] Assignment: lhs/rhs
- [x] Schedule: [int] operator
- [x] if_body/else_body return a Schedule node (not the list of their children)

sergisiso commented:
@arporter @rupertford This is ready for review, this PR simply adds the navigation methods for Schedule, Assignment and IfBlocks discussed in #411 . The Loop methods will be introduced in #400 .

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -569,6 +569,19 @@ def test_sched_view(capsys):
+
+    sched = psy.invokes.invoke_list[0].schedule
+    for indx in range(len(sched._children)):
+        assert sched[indx] is sched._children[indx]
```
What happens with an out-of-bounds request? Would be good to test that it does what you expect.

sergisiso replied:
Added out-of-bounds test. Also range-indexing (`sched[:]`) test as it is also an option.

arporter replied:
Great.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -569,6 +569,19 @@ def test_sched_view(capsys):
 
 
+def test_sched_getitem():
+    '''Test that Schedule have the [int] operators overloaded to return the
```
Picky but s/have/has/ and s/operators/operator/

sergisiso replied:
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1574,6 +1574,17 @@ def coloured_text(self):
 
+    def __getitem__(self, index):
+        '''
+        Overload the subscript notation ([int]) to access a Schedule statements
```
I decided to rephrase to "to access specific statements in the Schedule."

arporter replied:
Thanks.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1574,6 +1574,17 @@ def coloured_text(self):
+        Overload the subscript notation ([int]) to access a Schedule statements
+        sequence.
+
+        :param int index: index of the statment to access.
```
Fixed 'statement's.
Do you mean the rtype `Node`? Schedule statements can be anything: assignments, Ifs, Loops, ... so `Node` is the common ancestor.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4752,7 +4763,7 @@ def if_body(self):
                 "children, but found {0}.".format(len(self.children)))
 
-        return self._children[1]._children
+        return self._children[1]
```
Ups, I forgot updating this docstring. Fixed.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4763,7 +4774,7 @@ def else_body(self):
         '''
         if len(self._children) == 3:
-            return self._children[2]._children
+            return self._children[2]
```
Fixed docstring and returned None.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -6930,6 +6941,22 @@ class Assignment(Node):
+    @property
+    def lhs(self):
+        '''
+        :returns: the child node representing the assignment Left-Hand Side.
```
Done.

sergisiso commented:
@arporter It is ready for next review. I addressed the comments and added a new section in the Developer's Guide. I am not sure why travis does not generate the coverage graph for this PR.

arporter commented on doc/developer_guide/developers.rst:
```diff
@@ -266,6 +266,33 @@ provides the following common interface:
+
+To solve this issue some nodes also provide methods for semantic navigation:
+
+- Schedule: subscript operator for indexing access. (e.g. `sched[3]` or `sched[2:4]`)
```
"for indexing into the list of children" perhaps?

sergisiso replied:
used: "for indexing the statements inside the Schedule." I am trying to avoid 'children' to not mix things with the above methods.

arporter commented on doc/developer_guide/developers.rst:
```diff
@@ -266,6 +266,33 @@ provides the following common interface:
+To solve this issue some nodes also provide methods for semantic navigation:
+
+- Schedule: subscript operator for indexing access. (e.g. `sched[3]` or `sched[2:4]`)
+- Assignment: rhs and lhs properties.
```
Please mark these as code, i.e. `rhs` and `lhs` and again below for `condition`, `if_body`...

sergisiso replied:
Done

arporter commented on doc/developer_guide/developers.rst:
```diff
@@ -266,6 +266,33 @@ provides the following common interface:
+- Assignment: rhs and lhs properties.
+- IfBlocks: condition, if_body and else_body properties.
+
+This is the recommended methods to navigate the tree for heterogeneous operations
```
s/This is/These are/ and "...the tree for analysis or operations that depend on the Node type."

sergisiso replied:
Done

arporter commented on doc/developer_guide/developers.rst:
```diff
@@ -266,6 +266,33 @@ provides the following common interface:
+This is the recommended methods to navigate the tree for heterogeneous operations
+or analysis.
+
+Additionally, the `walk` method (available in all nodes) is able to recurse
```
Would be good to provide its interface here using autodoc (or maybe provide a link to its description in the reference guide).

sergisiso replied:
Added the autodoc to keep it coherent with the rest of the Developer's Guide. (but having links to the reference guide and a more lean developer's guide is something we could do in the future.)

sergisiso commented:
@arporter I updated the documentation. Ready for next review.
